File: 2022/14/advent202214.js
const fs = require('fs');

let input = fs.readFileSync('input202214.txt', 'utf-8', (err, data) => {
  if (err) {
    console.log(err)
  } else {
    return data;
  }
});

input = input.split('\n');
let maxy = 0;
let minx = Infinity;
let maxx = 0;
const waterfall = new Set();
for (const line of input) {
  const parse = line.split(' -> ').map((a) => a.split(',').map(Number));
  for (const coord of parse) {
    maxy = Math.max(maxy, coord[1]);
    minx = Math.min(minx, coord[0]);
    maxx = Math.max(maxx, coord[0]);
  }
  for (let i = 1; i < parse.length; i++) {
    if (parse[i][0] < parse[i - 1][0]) {
      for (let x = parse[i][0]; x <= parse[i - 1][0]; x++) {
        const coord = x + ',' + parse[i][1];
        waterfall.add(coord);
      }
    } else if (parse[i][0] > parse[i - 1][0]) {
      for (let x = parse[i][0]; x >= parse[i - 1][0]; x--) {
        const coord = x + ',' + parse[i][1];
        waterfall.add(coord);
      }
    } else if (parse[i][1] < parse[i - 1][1]) {
      for (let y = parse[i][1]; y <= parse[i - 1][1]; y++) {
        const coord = parse[i][0] + ',' + y;
        waterfall.add(coord);
      }
    } else {
      for (let y = parse[i][1]; y >= parse[i - 1][1]; y--) {
        const coord = parse[i][0] + ',' + y;
        waterfall.add(coord);
      }
    }
  }
}

let sand = 0;
let rest = true;
while (rest) {
  let x = 500;
  let y = 0;
  rest = false;
  while (!rest && x >= minx && x <= maxx && y <= maxy) {
    y++;
    if (waterfall.has(x + ',' + y)) {
      x--;
      if (waterfall.has(x + ',' + y)) {
        x += 2;
        if (waterfall.has(x + ',' + y)) {
          waterfall.add((x - 1) + ',' + (y - 1));
          rest = true;
          sand++;
        }
      }
    }
  }
}

console.log(sand);

const ground = maxy + 2;

while (!waterfall.has('500,0')) {
  let x = 500;
  let y = 0;
  rest = false;
  while (!rest && y < ground) {
    y++;
    if (waterfall.has(x + ',' + y)) {
      x--;
      if (waterfall.has(x + ',' + y)) {
        x += 2;
        if (waterfall.has(x + ',' + y)) {
          waterfall.add((x - 1) + ',' + (y - 1));
          rest = true;
          sand++;
        }
      }
    }
  }
  if (y === ground) {
    waterfall.add(x + ',' + y);
  }
}

console.log(sand);

File: 2022/14/advent202214.py
import re

file1 = open('input.txt','r')
lines = file1.readlines()

rocks = set()
maxY = 0

for line in lines:
  nums = list(map(int, re.findall('\d+', line)))
  for i in range(0, len(nums) - 2, 2):
    x0 = nums[i]
    y0 = nums[i + 1]
    x1 = nums[i + 2]
    y1 = nums[i + 3]
    maxY = max(maxY, y0, y1)
    if x0 < x1:
      for i in range(x0, x1 + 1):
        rocks.add((i, y0))
    elif x1 < x0:
      for i in range(x1, x0 + 1):
        rocks.add((i, y0))
    elif y0 < y1:
      for i in range(y0, y1 + 1):
        rocks.add((x0, i))
    elif y1 < y0:
      for i in range(y1, y0 + 1):
        rocks.add((x0, i))

trail = [(500, 0)]

rest = False

def trailer():
  global trail
  x, y = trail[-1]
  rest = False
  while not rest and y < maxY + 1:
    y += 1
    if (x, y) not in rocks:
      trail.append((x, y))
    elif (x - 1, y) not in rocks:
      x -= 1
      trail.append((x, y))
    elif (x + 1, y) not in rocks:
      x += 1
      trail.append((x, y))
    else:
      rest = True
  return rest

sand = 0

while trailer():
  rocks.add(trail.pop())
  sand += 1

print(sand)

while len(trail):
  trailer()
  rocks.add(trail.pop())
  sand += 1

print(sand)





